File: python/53_combinatoric_selections.py
def combinatoric_selections() -> int:
    
    # nCr = nC(n - r)
    # if nCr > 1000000, nC(n + 1), nC(n + 2), ..., nC(n - r) will also be > 1000000
    # values in range [r, n - r]: (n - r) - r + 1 = n - 2r + 1
    # unique values of nCr: 1, 2, 3, ..., n - r + 1
    
    # nC1 = n
    # nC2 = n! / (2! * (n - 2)!) = n * (n - 1) / 2
    # nC3 = n! / (3! * (n - 3)!) = n * (n - 1) * (n - 2) / (2 * 3)
    # nCr = nC(r - 1) * (n - r + 1) / r
    
    count: int = 0
    
    # 23 choose 10 is first value exceeding 1 million for 1 <= n <= 100
    for n in range(23, 101):
        ncr = n
        
        for r in range(2, n + 1):
            ncr = ncr * (n - r + 1) // r
            
            if ncr > 1_000_000:
                count += n - 2*r + 1
                break
        
    return count

if __name__ == "__main__":
    print(combinatoric_selections())

File: java/_53_CombinatoricSelections.java
import java.util.*;

public class _53_CombinatoricSelections{
    
    public static void main(String[] args){
        
        int count = 0;
        
        for (int n = 1; n <= 100; n++){
            int combs = n; // n choose 1 = n
            
            for (int r = 2; r <= n - r + 1; r++){
                // nCr = nC(n - r), so all the unique values for a given value of n
                // will be found halfway through the possible values of r.
                // if nCr > 1 million, then nC(r+1), nC(r+2), ..., nC(n-r) will also be greater than 1 million.
                
                // The range [r, n-r] has n - 2r + 1 values, so when a value of r is found where
                // nCr exceeds 1 million, the count is incremented by n - 2r + 1 and the next value of n is checked.
                
                combs = ((n - r + 1) * combs / r);
                // System.out.printf("%dC%d = %d %b\n", n, r, combs, combs > 1e6);
                if (combs > 1e6){
                    // System.out.println(r + " " + (n - r) + " " + n + " " + (n - 2*r + 1));
                    count += n - 2*r + 1;
                    break;
                }
            }
            // System.out.println();
        }
        
        System.out.println("Number of values of nCr, 1 <= n <= 100, greater than 1 million: " + count); // 4075
    }
}